scheds/rust/scx_rustland/src/bpf/main.bpf.c

```diff
@@ -253,49 +262,112 @@ static bool test_and_clear_usersched_needed(void)
  * doesn't run concurrently with the user-space scheduler (that is single
  * threaded), therefore this check is also safe from a concurrency perspective.
  */
-static bool is_usersched_needed(void)
+static bool usersched_has_pending_tasks(void)
 {
 	return nr_queued || nr_scheduled;
 }
 
 /*
- * Dispatch a task and wake-up a target CPU.
+ * Return the corresponding CPU associated to a DSQ.
+ */
+static s32 dsq_id_to_cpu(u64 dsq_id)
+{
+	if (dsq_id >= MAX_CPUS) {
+		scx_bpf_error("Invalid dsq_id: %llu", dsq_id);
+		return -EINVAL;
+	}
+	return (s32)dsq_id;
+}
+
+/*
+ * Return the DSQ ID associated to a CPU, or SHARED_DSQ if the CPU is not
+ * valid.
  */
-static void
-dispatch_task(struct task_struct *p, s32 cpu, u64 enq_flags)
+static u64 cpu_to_dsq_id(s32 cpu)
+{
+	if (cpu < 0 || cpu >= MAX_CPUS) {
+		scx_bpf_error("Invalid cpu: %d", cpu);
+		return SHARED_DSQ;
+	}
+	return (u64)cpu;
+}
+
+/*
+ * Dispatch a task to a target DSQ, waking up the corresponding CPU, if needed.
+ */
+static void dispatch_task(struct task_struct *p, u64 dsq_id, u64 enq_flags)
 {
 	u64 slice = __sync_fetch_and_add(&effective_slice_ns, 0) ? : slice_ns;
+	s32 cpu;
 
-	/*
-	 * Wake-up the CPU selected by the scheduler, so the task will have
-	 * more chances to transition there.
-	 */
-	if (cpu >= 0)
+	switch (dsq_id) {
+	case SCX_DSQ_LOCAL:
+		break;
+	case SHARED_DSQ:
+		/*
+		 * Dispatch a task to the shared DSQ and kick the CPU assigned
+		 * to the task by the select_cpu() callbak.
+		 */
+		cpu = scx_bpf_task_cpu(p);
 		scx_bpf_kick_cpu(cpu, 0);
-	dbg_msg("dispatch: pid=%d (%s) cpu=%ld", p->pid, p->comm, cpu);
-	scx_bpf_dispatch(p, SCX_DSQ_GLOBAL, slice, enq_flags);
+		break;
+	default:
+		/*
+		 * Dispatch a task to a specific per-CPU DSQ if the target CPU
+		 * can be used (according to the cpumask), otherwise redirect
+		 * the task to the shared DSQ.
+		 *
+		 * In the future we may want to provide a way to check the
+		 * cpumask in advance from user-space in a proper synchronized
+		 * way, instead of bouncing the task somewhere else, but for
```

Comment:
So, this should only be a problem when cpus_allowed changes. That synchronization is performed through `.dequeue()` and the problem is that that path can't sleep, so implementing actual handshake with userland is likely not possible. The sequence of events is likely as follows:

1. Task gets enqueued and pushed to userspace and queued there.
2. The task's cpus_allowed is going to be updated, so the task gets dequeued (ignored by rustland) and the cpumask is being updated.
3. Userland sched pushes the task to the kernel for dispatching. Actual dispatch hasn't happened yet.
4. The task's cpus_allowed is updated and the task gets reenqueued.
5. The task gets dispatched targeting now wrong CPU.

So, in scenarios like this, it *should* be safe to just ignore the task. The task is already re-enqueued, so it will get re-dispatched. The problem with the earlier attempt likely was that it was dispatched to the wrong dsq. Then, later when the re-enqueued task gets dispatched, it's ignored as it's already on another dsq.

Reply:
Thanks tons for looking at this @htejun !

I've tried to ignore the task completely when the cpumask is not valid, but I still get the `failed to run for 5s` error. To reproduce this I'm always dispatching to CPU#2 from the user-space scheduler and then I run a `stress-ng --race-sched 16` that spawn 16 tasks that are constantly bouncing around using sched_setaffinity() in a busy loop.

However, it looks like we can actually ignore the task if target CPU is not valid *and* also if the CPU assigned to the task (scx_bpf_task_cpu()) is also not valid, see d13ed5c. Does it make sense to you? With this in place everything seems to work and I can't break the scheduler. So I added this change to the PR and updated some comments here and there.

I'm currently stress testing this, but so far so good.